rs/sns/governance/src/neuron.rs

```diff
@@ -243,75 +249,138 @@ impl Neuron {
+    pub(crate) fn vote_from_ballots_following(
         &self,
         function_id: u64,
+        topic: Topic,
         ballots: &BTreeMap<String, Ballot>,
-        proposal_criticality: ProposalCriticality,
+        proposal_id: &ProposalId,
     ) -> Vote {
         // Step 1: Who are the relevant followees?
 
-        let empty = vec![];
-        let get_followee_neuron_ids = |function_id| -> &Vec<NeuronId> {
-            self.followees
-                .get(&function_id)
-                .map(|followees_message| &followees_message.followees)
-                // If there was no Followees object, then result is empty Vec. Therefore, we treat
-                // None the same as Some(Followees { followees: vec![] }).
-                .unwrap_or(&empty)
-        };
-
-        let mut followee_neuron_ids = get_followee_neuron_ids(function_id);
-
-        // If the function is not critical, and this Neuron does not have followees specifically for
-        // the function, then fall back to the "catch-all" following.
-        if followee_neuron_ids.is_empty() {
-            use ProposalCriticality::{Critical, Normal};
-            match proposal_criticality {
-                Normal => {
-                    let fallback_pseudo_function_id = u64::from(&Action::Unspecified(Empty {}));
-                    followee_neuron_ids = get_followee_neuron_ids(fallback_pseudo_function_id);
-                }
-                Critical => (),
-            }
-        }
+        // Step 1.1. Try to get relevant followees from topic following, if available.
+        let followees = self.get_followee_neuron_ids(function_id, topic);
 
-        // This is needed to avoid returning No in Step 3 due to no followees.
-        if followee_neuron_ids.is_empty() {
+        if followees.is_empty() {
             return Vote::Unspecified;
         }
 
         // Step 2: Count followee votes.
+
         let mut yes: usize = 0;
         let mut no: usize = 0;
-        for followee_neuron_id in followee_neuron_ids {
-            let followee_vote = match ballots.get(&followee_neuron_id.to_string()) {
-                Some(ballot) => ballot.vote,
-                None => {
-                    // We are following someone who doesn't even have an empty
-                    // ballot. Maybe this followee should be removed?
-                    continue;
-                }
+        for neuron_id in &followees {
+            let Some(ballot) = ballots.get(&neuron_id.to_string()) else {
+                log!(
+                    ERROR,
+                    "Skipping followee neuron {} without a ballot for proposal {}",
+                    neuron_id,
+                    proposal_id.id,
+                );
+                continue;
+            };
+
+            let Ok(followee_vote) = Vote::try_from(ballot.vote) else {
+                log!(
+                    ERROR,
+                    "Skipping followee neuron {} with an invalid vote {} for proposal {}",
```

Comment:
Actually, this log is an error that you should never see.  So it makes sense.  You could even put a debug assert here b/c it should blow up during development if we ran across a vote being outside of the range of 0,1,2

Reply:
NCR on this one.